src/java/org/apache/cassandra/io/util/WrappingRebufferer.java

```diff
@@ -144,5 +145,11 @@ public void release()
 
             buffers.offerFirst(this);
         }
+
+        @VisibleForTesting
+        public void offset(long l)
```

Comment:
hmm... `WrappingRebufferer` in pre-DB-1220 version does not look right to me. Say:

```
bh1 = wr.rebuffer(1000)
bh2 = wr.rebuffer(10000)
bh1.release()
bh2.release()
```

We may actually obtain two different buffer holders from the underlying rebuffer. However `WrappingRebuffer.rebuffer` always returns itself as BufferHolder. Therefore, when releasing, we would release the last bh twice and never release the first bh as the reference to it is lost after creating the second one. I'm saying that this makes WR able to work on a single BH at a time and when I added assertion on that, the some test started to fail.

Reply:
Which test is failing?

I still prefer to make it clear that this access pattern is forbidden and add the assertion. We should add a comment (both here and in `TailOverridingRebufferer`) explaining it, something along the lines of the javadoc on `LimitingRebufferer`.

This pattern (non-multiple-access-safe) is used for all reader-specific rebufferers, including `LimitingRebufferer` and the two variations of `BufferManagingRebufferer`; if we make this safe, we should also change them, and we should seriously consider making all thread-safe as well.

In any case, the code in `RandomAccessReader.rebufferAt` can cause a double release when an exception is thrown in `rebufferer.rebuffer`. We should fix this by inserting
```
        bufferHolder = Rebufferer.EMPTY;    // to avoid double close if the call below throws
```
after the release.

Reply:
I can see the reason for the test failures: `Walker.go` is also using the TPC-imposed access pattern. The same solution as the above can be applied:
```
diff --git a/src/java/org/apache/cassandra/io/tries/Walker.java b/src/java/org/apache/cassandra/io/tries/Walker.java
--- a/src/java/org/apache/cassandra/io/tries/Walker.java	(revision a25cc0507a87eb41934fd8917c6d7b6797fad8c6)
+++ b/src/java/org/apache/cassandra/io/tries/Walker.java	(date 1618473141380)
@@ -83,9 +83,9 @@
         long curOffset = position - bh.offset();
         if (curOffset < 0 || curOffset >= buf.limit())
         {
-            BufferHolder currentBh = bh;
+            bh.release();
+            bh = Rebufferer.EMPTY; // to avoid double close if the call below throws
             bh = source.rebuffer(position);
-            currentBh.release();
             buf = bh.buffer();
             curOffset = position - bh.offset();
             assert curOffset >= 0 && curOffset < buf.limit() : String.format("Invalid offset: %d, buf: %s, bh: %s", curOffset, buf, bh);
```

Reply:
exactly that

Reply:
When both chunk cache and memory mapping are disabled, the current `Walker.go` code is an invalid pattern for `BufferManagingRebufferer` too. We don't see an error from that because releases are no-ops, but I think we should add a no-rebuffer-before-release assertion there too.
`LimitingRebufferer` can be a descendant of `WrappingRebufferer` -- we can copy the DSE version of it.